File: MasteringDart/lib/chapters/chapter10/basic/basic_null_safety.dart
// ignore_for_file: avoid_print
//
//  basic_null_safety.dart
//  MasteringDart
//
//  Capítulo 10 — Seguridad nula (Null Safety).
//  Nivel BÁSICO: Variables nullable, comprobaciones null, valores por defecto.
//
//  Dart: los tipos nullable usan ? (ej. String?); null es el valor.
//
//  Ejecutar con: dart run lib/chapters/chapter10/basic/basic_null_safety.dart

void runBasicNullSafetyExamples() {
  // --- String nullable: tipo String? (puede contener null) ---
  String? name = 'Alice';
  String? nickname = null;

  // --- Comprobar si el valor es null ---
  if (name != null) {
    print('Name length: ${name.length}');
  }
  if (nickname == null) {
    print('Nickname is null');
  }

  // --- Valores por defecto: operador ?? ---
  final displayName = name ?? 'Unknown';
  final displayNick = nickname ?? 'No nickname';
  print('Display name: $displayName');
  print('Display nick: $displayNick');

  // --- Acceso seguro: ?. devuelve null en lugar de lanzar ---
  print('Name length (safe): ${name?.length}');
  print('Nickname length (safe): ${nickname?.length}');
}

void main() {
  runBasicNullSafetyExamples();
}

File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter10/basic/BasicNullSafety.kt
package com.devlokos.masteringkotlin.chapters.chapter10.basic

/**
 * Capítulo 10 — Null Safety.
 * Nivel BÁSICO: Variables nullable, comprobaciones de null, valores por defecto.
 *
 * Kotlin: los tipos nullable usan ? (ej. String?)
 */
fun main() {
    runBasicNullSafetyExamples()
}

fun runBasicNullSafetyExamples() {
    // --- String nullable: el tipo es String? (puede contener null) ---
    val name: String? = "Alice"
    val nickname: String? = null

    // --- Comprobar si el valor es null ---
    if (name != null) {
        println("Name length: ${name.length}")
    }
    if (nickname == null) {
        println("Nickname is null")
    }

    // --- Imprimir valores por defecto: operador Elvis ?: ---
    val displayName = name ?: "Unknown"
    val displayNick = nickname ?: "No nickname"
    println("Display name: $displayName")
    println("Display nick: $displayNick")

    // --- Safe call: ?. devuelve null en lugar de lanzar ---
    println("Name length (safe): ${name?.length}")
    println("Nickname length (safe): ${nickname?.length}")
}
